File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    // min_heap for all n/2 (second half of the array)
    // max_heap for all n/2+1 element(first half);
    // if n is even -> min_heap+max_heap / 2 
    // if n is odd -> max_heap top element
    priority_queue<int,vector<int>,greater<int>> minh;
    priority_queue<int> maxh;
    int cnt;
    MedianFinder() {
        this->cnt=0;
    }
    
    void addNum(int num) {
        ++cnt;
        minh.push(num);
        if(cnt&1){
            maxh.push(minh.top());
            minh.pop();
        }
        if(cnt&1==0){
            while(minh.size()!=maxh.size()){
                if(minh.size()>maxh.size()){
                    maxh.push(minh.top());
                    minh.pop();
                }
                else{
                    minh.push(maxh.top());
                    maxh.pop();
                }
            }
        }
        else{
            while(maxh.size()<(cnt/2+1)){
                maxh.push(minh.top());
                minh.pop();
            }
            while(minh.size()<cnt/2){
                minh.push(maxh.top());
                maxh.pop();
            }

        }
    }
    
    double findMedian() {
        if(cnt&1) return (double)maxh.top();
        double ans = (double)minh.top()+(double)(maxh.top());
        return ans/2.0;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
from heapq import heapify,heappush,heappop 

class MedianFinder:

    def __init__(self):
        self.min_heap =[] # right half 
        self.max_heap =[] # left half

    def resetHeap(self,max_heap,min_heap): 
        # we push max_heap 
        # if total is even -> ensure both heaps are same size 
        # if total is odd -> ensure size(max_heap)-size(min_heap)==1
        total = len(max_heap) + len(min_heap)
        if total%2: 
            # ensure max_heap > 1 min_heap in size 
            while len(max_heap)!=1+len(min_heap):
                if len(max_heap)>len(min_heap):
                    heappush(min_heap,abs(heappop(max_heap)))
                else: 
                    heappush(max_heap,-heappop(min_heap))
        else: 
            while len(max_heap)!=len(min_heap): 
                if len(max_heap)>len(min_heap):
                    heappush(min_heap,-(heappop(max_heap)))
                else: 
                    heappush(max_heap,-heappop(min_heap))


    def addNum(self, num: int) -> None:
        # place the value in correct heap
        if len(self.min_heap) and self.min_heap[0]<num:
            heappush(self.min_heap,num) 
        elif len(self.max_heap) and -(self.max_heap[0])>num:
            heappush(self.max_heap,-num)
        else: 
            heappush(self.max_heap,-num)

        self.resetHeap(max_heap=self.max_heap,min_heap=self.min_heap)

    def findMedian(self) -> float:
        total = len(self.min_heap)+len(self.max_heap) 
        if total%2: 
            return -(self.max_heap[0])
        else:
            return (-self.max_heap[0]+self.min_heap[0])/2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()